File: webgl/source/js/gl/shaders/vertexshaderbuilder.js
class VertexShaderBuilder extends ShaderBuilderA {
    mLightAmbient
    mLightDirectionalColor
    mLightDirectionalVector

    directionalLightEmbeded(
        ambient, // float array
        lightDirectionalColor, // float array
        lightDirectionalVector // float array
    ) {
        this.mLightAmbient = ambient;
        this.mLightDirectionalColor = lightDirectionalColor;
        this.mLightDirectionalVector = lightDirectionalVector;
        return this;
    }

    buildCalcLightEmbeded() {
        if (this.mLightAmbient == null ||
            this.mLightDirectionalColor == null ||
            this.mLightDirectionalVector == null) {
            return "";
        }
        var sb = "";
        sb = sb + `vec3 lightAmbient = vec3(${this.mLightAmbient[0]}, ${this.mLightAmbient[1]}, ${this.mLightAmbient[2]});\n `;
        sb = sb + `vec3 lightDirectionalColor = vec3(${this.mLightDirectionalColor[0]}, ${this.mLightDirectionalColor[1]}, ${this.mLightDirectionalColor[2]});\n `;
        sb = sb + `vec3 lightDirectionalVector = vec3(${this.mLightDirectionalVector[0]}, ${this.mLightDirectionalVector[1]}, ${this.mLightDirectionalVector[2]});\n `;
        sb = sb + "vec4 transformedNormal = normalMatrix*vec4(vNormal,1.0);\n";
        sb = sb + "float directional = max(dot(transformedNormal.xyz, lightDirectionalVector), 0.0);\n";
        sb = sb + "exLighting =  lightAmbient + (lightDirectionalColor * directional);";

        return sb;
    }

    exchangeData(color, texCoord, lighting) {
        this.setExchangeData(color, texCoord, lighting);
        return this;
    }

    matrixData(normal) {
        this.mMatrix = true;
        this.mNormalMatrix = normal;
        this.mModelMatrix = false;
        this.mViewMatrix = false;
        this.mViewModelMatrix = false;
        this.mProjectionMatrix = false;
        return this;
    }

    matrixViewModelData(normal) {
        this.mMatrix = false;
        this.mNormalMatrix = normal;
        this.mModelMatrix = false;
        this.mViewMatrix = false;
        this.mViewModelMatrix = true;
        this.mProjectionMatrix = true;
        return this;
    }

    matrix3Data(normal) {
        this.mMatrix = false;
        this.mNormalMatrix = normal;
        this.mModelMatrix = true;
        this.mViewMatrix = true;
        this.mViewModelMatrix = false;
        this.mProjectionMatrix = true;
        return this;
    }

    vertexInputData(coord, color, texCoord, normal) {
        this.setVertexInputData(coord, color, texCoord, normal)
        return this;
    }

    version(versionValue) {
        this.setVersion(versionValue);
        return this;
    }

    precision(val) {
        this.setPrecision(val);
        return this;
    }

    buildCalcCustom() {
        var sb = "";
        sb = sb + this.buildCalcPosition();
        sb = sb + this.buildCalcTexCoord();
        sb = sb + this.buildCalcColor();
        sb = sb + this.buildCalcLightEmbeded();
        return sb;
    }
}

File: lwjgl/src/main/java/com/darkraha/opengldemoj/gl/shaders/VertexShaderBuilder.java
package com.darkraha.opengldemoj.gl.shaders;

public class VertexShaderBuilder extends ShaderBuilderA {

    private float[] mLightAmbient;
    private float[] mLightDirectionalColor;
    private float[] mLightDirectionalVector;


    public VertexShaderBuilder directionalLightEmbeded(float[] ambient,
                                                       float[] lightDirectionalColor,
                                                       float[] lightDirectionalVector
    ) {
        mLightAmbient = ambient;
        mLightDirectionalColor = lightDirectionalColor;
        mLightDirectionalVector = lightDirectionalVector;
        return this;
    }


    protected String buildCalcLightEmbeded() {
        if (mLightAmbient == null || mLightDirectionalColor == null
                || mLightDirectionalVector == null) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        sb.append("vec3 lightAmbient = vec3(")
                .append(mLightAmbient[0]).append(",")
                .append(mLightAmbient[1]).append(",")
                .append(mLightAmbient[2]).append(");\n");

        sb.append("vec3 lightDirectionalColor = vec3(")
                .append(mLightDirectionalColor[0]).append(",")
                .append(mLightDirectionalColor[1]).append(",")
                .append(mLightDirectionalColor[2]).append(");\n");

        sb.append("vec3 lightDirectionalVector = vec3(")
                .append(mLightDirectionalVector[0]).append(",")
                .append(mLightDirectionalVector[1]).append(",")
                .append(mLightDirectionalVector[2]).append(");\n");


        sb.append("vec4 transformedNormal = normalMatrix*vec4(vNormal,1.0);\n");
        sb.append("float directional = max(dot(transformedNormal.xyz, lightDirectionalVector), 0.0);\n");
        sb.append("exLighting =  lightAmbient + (lightDirectionalColor * directional);");
        return sb.toString();
    }


    public VertexShaderBuilder exchangeData(boolean color, boolean texCoord, boolean lighting) {
        setExchangeData(color, texCoord, lighting);
        return this;
    }

    public VertexShaderBuilder matrixData(boolean normal) {
        mMatrix = true;
        mNormalMatrix = normal;
        mModelMatrix = false;
        mViewMatrix = false;
        mViewModelMatrix = false;
        mProjectionMatrix = false;
        return this;
    }

    public VertexShaderBuilder matrixViewModelData(boolean normal) {
        mMatrix = false;
        mNormalMatrix = normal;
        mModelMatrix = false;
        mViewMatrix = false;
        mViewModelMatrix = true;
        mProjectionMatrix = true;
        return this;
    }


    public VertexShaderBuilder matrix3Data(boolean normal) {
        mMatrix = false;
        mNormalMatrix = normal;
        mModelMatrix = true;
        mViewMatrix = true;
        mViewModelMatrix = false;
        mProjectionMatrix = true;
        return this;
    }


    public VertexShaderBuilder vertexInputData(boolean coord, boolean color, boolean texCoord, boolean normal) {
        setVertexInputData(coord, color, texCoord, normal);
        return this;
    }

    public VertexShaderBuilder version(String versionValue) {
        setVersion(versionValue);
        return this;
    }

    public VertexShaderBuilder precision(int val) {
        setPrecision(val);
        return this;
    }


    @Override
    protected String buildCalcCustom() {
        StringBuilder sb = new StringBuilder();
        sb.append(buildCalcPosition());
        sb.append(buildCalcTexCoord());
        sb.append(buildCalcColor());
        sb.append(buildCalcLightEmbeded());
        return sb.toString();
    }
}
